tiledb/common/evaluator/evaluator.h

```diff
@@ -0,0 +1,249 @@
+   * recently used element from the cache.
+   */
+  List lru_;
+
+  /**
+   * Each entry stores a promised value which is in the process of
+   * being fetched. Readers will wait on it to become available.
+   */
+  std::unordered_map<Key, OndemandValue> inprogress_;
+
+  /**
+   * Map of cached entries by key. Each entry stores and iterator
+   * within the LRU list and a shared_ptr to the value.
+   * If an entry is in this map, it is guaranteed that the value is
+   * already available to serve readers.
+   */
+  std::unordered_map<Key, Cached_Entry> entries_;
+
+  /** Bookkeeping lock */
+  std::mutex mutex_;
+};
+
+template <class Key, class Value>
+class NoCachePolicy {
+ public:
+  using key_type = Key;
+  using value_type = Value;
+
+  NoCachePolicy() = default;
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+};
+
+template <class Key, class Value, size_t N = 0>
+class MaxEntriesCache : public CachePolicyBase<Key, Value> {
+ public:
+  MaxEntriesCache()
+      : num_entries_(0) {
+  }
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+
+  virtual std::shared_ptr<Value> enforce_policy(
+      [[maybe_unused]] size_t mem_usage) override {
+    if (num_entries_ == max_entries_) {
+      auto evicted_value = this->evict();
+      --num_entries_;
+      return evicted_value;
+    }
+    ++num_entries_;
+    return nullptr;
+  }
+
+ private:
+  size_t num_entries_;
+  static const size_t max_entries_ = N;
+};
+
+template <class Key, class Value, size_t N = 1024>
+class MemoryBudgetedCache : public CachePolicyBase<Key, Value> {
+ public:
+  MemoryBudgetedCache()
+      : memory_consumed_(0) {
+  }
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+
+  virtual std::shared_ptr<Value> enforce_policy(size_t mem_usage) override {
+    if (memory_consumed_ + mem_usage > memory_budget_) {
+      auto evicted_value = this->evict();
+      memory_consumed_ -= mem_usage;
+      return evicted_value;
+    }
+
+    memory_consumed_ += mem_usage;
+    return nullptr;
+  }
+
+ private:
+  /** The maximum amount of bytes managed by this cache */
+  static const size_t memory_budget_ = N;
+
+  /** The maximum amount of bytes currently consumed by this cache */
+  size_t memory_consumed_;
+};
+
+template <class Policy, class F>
+class Evaluator {
+ public:
+  using Key = typename Policy::key_type;
+  using Value = typename Policy::value_type;
+
+  Evaluator()
+      : caching_policy_{} {
+  }
+
+  /**
+   * Evaluate callback on key and return the value.
+   * According to the policy specified, the value might be fetched
+   * directly from the cache or the callback might be executed to fetch it.
+   */
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return caching_policy_.template operator()<F>(f, key);
+  }
```

Comment:
`operator()` needs to be of the form `return_type operator()(Key key)`. The way it's written now, it would more appropriately be called `lambda`. There are two other ways of not having it as a function argument here.

1. It could be a template argument. Taking a callable as a template argument needs to be in two declaration steps, the first one starting with `auto f`. See the exception wrapper for how to do the declaration. In that case its a function template, but the argument deduction rules are the same for class templates.
2. It could be a constructor argument.

Which one to pick is somewhat arbitrary, but really depends on whether the function is parametric, and thus needs parameters at construction time, or is not parametric, and can be declared statically. For now I'd go with a constructor argument; it's more general and a bit easier to declare. It has more overhead, but it's not large enough to worry about at this stage.